events/controller/controller.go

```diff
@@ -0,0 +1,483 @@
+
+	errListenEvents := errors.New("listen events error")
+	if chf == nil {
+		return errors.Wrap(errListenEvents, "expected valid ConditionHandlerFactor, got nil")
+	}
+
+	if n.stream == nil {
+		return errors.Wrap(errListenEvents, "connection not initialized, invoke Connect() before calling this method")
+	}
+
+	n.conditionHandlerFactory = chf
+
+	pullTicker := time.NewTicker(n.pullEventInterval)
+	defer pullTicker.Stop()
+
+Loop:
+	for {
+		select {
+		case <-pullTicker.C:
+			if n.concurrencyLimit() {
+				continue
+			}
+
+			if err := n.processEvents(ctx); err != nil {
+				return errors.Wrap(errListenEvents, err.Error())
+			}
+
+		case <-ctx.Done():
+			if n.dispatched > 0 {
+				continue
+			}
+
+			break Loop
+		}
+	}
+
+	return nil
+}
+
+// process event into a condition
+func (n *NatsController) processEvents(ctx context.Context) error {
+	pullCtx, cancel := context.WithTimeout(ctx, n.pullEventTimeout)
+	defer cancel()
+
+	errProcessEvent := errors.New("process event error")
+	msgs, err := n.stream.PullMsg(pullCtx, 1)
+	switch {
+	case err == nil:
+	case errors.Is(err, nats.ErrTimeout):
+		n.logger.WithFields(
+			logrus.Fields{"info": err.Error()},
+		).Trace("no new events")
+		return nil
+
+	default:
+		n.logger.WithFields(
+			logrus.Fields{"err": err.Error()},
+		).Warn("retrieving new messages")
+
+		metricsNATSError("pull-msg")
+		return errors.Wrap(errProcessEvent, err.Error())
+	}
+
+	for _, msg := range msgs {
+		if n.concurrencyLimit() {
+			return nil
+		}
+
+		// event status setter to keep the JS updated on our progress
+		eventAcknowleger := n.newNatsEventStatusAcknowleger(msg)
+
+		if ctx.Err() != nil {
+			eventAcknowleger.nak()
+
+			return errors.Wrap(errProcessEvent, ctx.Err().Error())
+		}
+
+		// setup the handle to query condition status
+		conditionStatusQueryor, err := n.NewNatsConditionStatusQueryor()
+		if err != nil {
+			eventAcknowleger.nak()
+
+			return errors.Wrap(errProcessEvent, ctx.Err().Error())
+		}
+
+		// spawn msg process handler
+		n.syncWG.Add(1)
+		go n.processConditionFromEvent(ctx, msg, eventAcknowleger, conditionStatusQueryor)
+	}
+
+	return nil
+}
+
+func (n *NatsController) processConditionFromEvent(ctx context.Context, msg events.Message, eventAcknowleger eventStatusAcknowleger, conditionStatusQueryor ConditionStatusQueryor) {
+	defer n.syncWG.Done()
+	ctx, span := otel.Tracer(pkgName).Start(
+		ctx,
+		"processConditionFromEvent",
+	)
+
+	atomic.AddInt32(&n.dispatched, 1)
+	defer atomic.AddInt32(&n.dispatched, -1)
+
+	cond, err := conditionFromEvent(msg)
+	if err != nil {
+		n.logger.WithError(err).WithField(
+			"subject", msg.Subject()).Warn("unable to retrieve condition from message")
+
+		metricsEventsCounter(false, "ack")
+		eventAcknowleger.complete()
+
+		return
+	}
+
+	// extract parent trace context from the event if any.
+	ctx = msg.ExtractOtelTraceContext(ctx)
+
+	conditionStatusPublisher, err := n.NewNatsConditionStatusPublisher(cond.ID.String())
+	if err != nil {
```

Comment:
If the controller cannot connect to the KV to publish status, it should fail earlier than later. For aesthetics - explained in a comment below - `processCondition` accepts all objects it needs to do its work instead of initialising them itself.